Introduce next-generation Python bindings using `nanobind`

ManifoldFR opened the pull request:
This (draft) pull request adds a second set of Python bindings using the modern [nanobind](https://github.com/wjakob/nanobind) bindings generator.

The new nanobind-based API can be imported as a replacement for the old one by importing:
```python
import coal
```

Only **one** of the bindings can be built at one time. The nanobind bindings can be built by setting the new `COAL_PYTHON_NANOBIND` option to ON.
```bash
cd <build-dir>
cmake . -DCOAL_PYTHON_NANOBIND=ON
```

~~These new bindings depend, at runtime, on the new [nanoeigenpy](https://github.com/Simple-Robotics/nanoeigenpy) support library, which we use to expose Eigen's Quaternion type.~~

# Task checklist 

- [x] Ensure no collision with previous set of Boost.Python bindings:
  - [x] cmake: Move Boost.Python bindings' code from main `CMakeLists.txt` to legacy `python/CMakeLists.txt`
  - [x] cmake: add flag `COAL_PYTHON_BINDINGS`
- [x] Stubs
- [x] Re-expose the following features from the Boost.Python bindings:
  - [x] Version
  - [x] Broadphase (`python/broadphase/broadphase.cc`)
  - [x] GJK (`python/gjk.cc`)
  - [x] MeshLoader
  - [x] Maths (`python/math.cc`)
  - [x] Pickling
  - [x] Serialization
  - [x] Contact patches
  - [x] Distance API (`python/distance.cc`)
  - [x] Octree (`python/octree.cc`)
  - [x] Collision (`python/collision.cc`)
  - [x] Collision Geometries (`python/collision-geometries.cc`)
    - [x] Base class
    - [x] BVH models
    - [x] Heightfields
    - [x] `CollisionObject`
- [x] Test everything

ManifoldFR committed:
pre-commit : defer to .clang-format file

ManifoldFR committed:
Add stump nanobind bindings file python-nb/coal.cc

ManifoldFR committed:
[python-nb] Add exposeVersion()

ManifoldFR committed:
python-nb : send nanobind module coal_pywrap_v2 to new Python module "coal.v2"

ManifoldFR committed:
Fix stub generation

florent-lamiraux commented:
For your information, after a quick test based on https://stackoverflow.com/questions/8233252/boostpython-and-weak-ptr-stuff-disappearing, nanobind seems as bad as boost python to handle weak pointers.

ManifoldFR commented:
> For your information, after a quick test based on https://stackoverflow.com/questions/8233252/boostpython-and-weak-ptr-stuff-disappearing, nanobind seems as bad as boost python to handle weak pointers.

Thanks, this is good to know in the future. Almost certainly, weak references/pointers will be an object management nightmare whatever the bindings generator is.

Anyhow, there is a definite advantage of upgrading our bindings, namely simplifying maintenance, getting modern features (and up-to-date docs), ability to use the Python 3.12+ stable ABI - which are things Boost.Python simply doesn't have and never will 😃

ManifoldFR committed:
python-nb : rename coal.cc to module.cc

ManifoldFR committed:
python-nb : Start work on math module

Add more to Transform3s binding

ManifoldFR committed:
pixi.toml : add nanobind dependency

ManifoldFR committed:
[python-nb] add serializable.hh

ManifoldFR committed:
Expose Mesh loaders and collision geometries' enums

ManifoldFR committed:
Expose CollisionObject

ManifoldFR committed:
Update CHANGELOG

ManifoldFR committed:
python-nb : Add pickle.hh, implement boost archive pickling

ManifoldFR committed:
python-nb : Expose BVH collision geoms

ManifoldFR committed:
Fix after rebase

ManifoldFR committed:
Make building Nanobind bindings an "or" rather and an "either"

ManifoldFR commented:
Change of plans: following discussions with @jorisv and @nim65s, the new bindings shall *not* coexist with the legacy Boost.Python bindings. Users will be able to build/install either one, but not both.

ManifoldFR committed:
fix legacy Boost.Python cmake listfile

ManifoldFR committed:
[cmake] make nanobind the default bindings

ManifoldFR committed:
Fix detection of nanobind, fix Pixi support, update CI, fix test/python_unit/CMakeLists.txt

ManifoldFR committed:
Move nanoeigenpy import

ManifoldFR committed:
python-nb/CMakeLists.txt : small change, remove MODULE_NAME var

ManifoldFR committed:
tiny update

Lucas-Haubert committed:
To debug: At complile time (No module named 'coal.coal_pywrap_nb') and at run time with tests (No module named [module] in coal)]

Lucas-Haubert committed:
Fixing the bug: exposes not done well

ManifoldFR committed:
python-nb/CMakeLists.txt : add missing install target  for nanobind-based extension module

ManifoldFR committed:
python-nb : fix installation of stub file

Lucas-Haubert committed:
re-fix generating stub -- generate stub for extension module (not the __init__)

Lucas-Haubert committed:
Remove includes to Boost.Python in python-nb

Lucas-Haubert committed:
Fix tests api.py and collision.py

Lucas-Haubert committed:
Fixed installation bug with PickleVisitor and SerializableVisitor

Lucas-Haubert committed:
Added copyright

Lucas-Haubert committed:
Removed copy pasted wrappers from boost bindings and coded with lambda functions instead

Lucas-Haubert committed:
Clean the includes

Lucas-Haubert committed:
Moved expose of AABB in separate file

Lucas-Haubert committed:
Expose the StdVec

Lucas-Haubert committed:
Put internal references for the bp with_custodian_and_ward_postcall

Lucas-Haubert committed:
updated return value policy

Lucas-Haubert committed:
Management of shared pointers via the include of <nanobind/stl/shared_ptr.h>

Lucas-Haubert committed:
Fixed test geometric_shapes with None as last argument and pre commit

Lucas-Haubert committed:
Debugging pickle with Transform3s passing but not other classes

ManifoldFR commented:
Commits from @Lucas-Haubert got merged in :)

ManifoldFR committed:
module.cc : undupe call to exposeOctree()

ManifoldFR committed:
python-nb/pickle.hh : placement-new a default-initialized object

ManifoldFR committed:
test/python_unit : re-enable all pickling tests

ManifoldFR commented on python-nb/pickle.hh:
```diff
@@ -0,0 +1,34 @@
+      }).def("__setstate__", [](T &obj, const std::string &state) {
+      std::istringstream is(state);
+      boost::archive::text_iarchive ia(is, boost::archive::no_codecvt);
+      new (&obj) T();
```
@jcarpent this line was required to avoid a memory read bug. It seems like a difference with Boost.Python is that nanobind does _not_ allocate the object passed to `__setstate__`, so we have to placement-new it.

I don't feel comfortable with placement-new'ing using the default constructor because it might not always exist. It works for now because all the objects we allow pickling for have default constructors. Do you know if Boost.Serialization could handle object construction for us?

ManifoldFR committed:
python-nb/shapes.cc : run formatting

ManifoldFR committed:
pixi.toml : add Python version guide

ManifoldFR committed:
nanobind stubgen requires typing_extensions for python-oldest env

Lucas-Haubert commented:
Hello @nim65s . The first batch of commits is done here. After discussing with @ManifoldFR, we wondered if you could adapt the pip workflow regarding the intergration of nanobind.

nim65s commented:
sure

ManifoldFR commented:
I've just rebased on devel.

ManifoldFR commented:
Actually rebasing on `devel` failed because of recent (major?) changes to the core Coal library. @Lucas-Haubert can you take a look?

Lucas-Haubert committed:
Fix compile bug with nanobing regarding the Convex16 and Convex32 PR

Lucas-Haubert committed:
Added remaining functions in octree and math

ManifoldFR committed:
macos-linux-pip.yml -- set to use Boost.Python bindings by default

Lucas-Haubert committed:
Fix CI pixi: Removed typename in static_cast

ManifoldFR committed:
python-nb/math.cc : simplify and fix

ManifoldFR committed:
Minor changes - fix path to Python extension module for tests

ManifoldFR committed:
python/CMakeLists.txt : remove superfluous message()

ManifoldFR committed:
debug push

ManifoldFR committed:
[boost-python bindings] remove `FINDPYTHON` call

ManifoldFR committed:
pip workflow : pass Python_EXECUTABLE

ManifoldFR committed:
[cmake] move scope of Python-finding code

ManifoldFR committed:
python-nb : add windows_dll_manager

ManifoldFR committed:
Use activation scripts to set Python executable path

ManifoldFR commented:
Oh my God I hate Windows so much

ManifoldFR committed:
Fix deprecation warning

ManifoldFR committed:
swap includes

jorisv committed:
cmake: Fix jrl-cmakemodules python unit test when using nanobind

jorisv committed:
cmake: Fix Windows nanobind stub generation

jorisv committed:
nanobind: Fix CL+nanobind+Boost.Serialization build issue

ManifoldFR commented on python-nb/fwd.h:
@Lucas-Haubert we forgot to put include guards here.. and other headers for the nanobind bindings as mentioned by the great @jorisv

ManifoldFR commented:
Only thing left is to fix the ROS build.

@nim65s would you have time for that, this week?

Lucas-Haubert committed:
nanobind: Added include guards in headers of nanobind python bindings

nim65s committed:
ROS: add nanobind dependency

fix:
> -- Found Python: /usr/bin/python3 (found version "3.12.3") found components: Interpreter Development.Module
> /usr/bin/python3: No module named nanobind
> CMake Error at python-nb/CMakeLists.txt:12 (find_package):
>   Could not find a package configuration file provided by "nanobind" with any
>   of the following names:
>
>     nanobindConfig.cmake
>     nanobind-config.cmake
>
>   Add the installation prefix of "nanobind" to CMAKE_PREFIX_PATH or set
>   "nanobind_DIR" to a directory containing one of the above files.  If
>   "nanobind" provides a separate development package or SDK, be sure it has
>   been installed.

nim65s committed:
nix: nanoeigenpy

nim65s committed:
ROS: add nanoeigenpy dependency

Not sure when this will land. The first step is waiting for
https://github.com/ros/rosdistro/pull/45744

ManifoldFR commented:
I think that, while we are waiting https://github.com/ros/rosdistro/pull/45744 to be merged, we can still merge this PR.

Further steps, in other PRs:
- port doxygen autodoc ?
- write-up a porting guide detailing our experience, perhaps some benchmarks, pursuant to https://github.com/wjakob/nanobind/discussions/1001

ManifoldFR committed:
pixi : add nanoeigenpy dependency

ManifoldFR committed:
pixi.toml : replace "project" by "workspace"
reset pixi.lock

ManifoldFR committed:
python-nb : expose functional (std::function-based) API for broadphase callbacks

ManifoldFR committed:
Update pixi lockfile

ManifoldFR committed:
Merge branch 'devel' into topic/nanobind

ManifoldFR commented:
With this our work is done. Only two things are the docstrings for the bindings (will be done later), and the ROS CI which is snafu

Ping @nim65s @jorisv

nim65s committed:
nix: test bp+nb

nim65s committed:
CMake: fix syntax

nim65s commented on CMakeLists.txt:
```diff
@@ -162,12 +162,22 @@ APPLY_DEFAULT_APPLE_CONFIGURATION()
-  OFF
+  COAL_PYTHON_NANOBIND
+  "Build the nanobind-based bindings."
+  ON
```
```suggestion
  OFF
```

Activating this by default would be an unnecessary breaking change. We can do that on the next major release.

nim65s commented on CMakeLists.txt:
```diff
@@ -512,7 +513,28 @@ endif(COAL_HAS_OCTOMAP)
+    PROPERTIES EXCLUDE_FROM_DEFAULT_BUILD True
+  )
+  if(COAL_PYTHON_NANOBIND)
+    if(CMAKE_VERSION VERSION_LESS 3.18)
+      set(DEV_MODULE Development)
+    else()
+      set(DEV_MODULE Development.Module)
+    endif()
+    find_package(Python REQUIRED COMPONENTS Interpreter ${DEV_MODULE})
```
```suggestion
    find_package(Python REQUIRED COMPONENTS Interpreter Development.Module)
```



we can target CMake 3.22, ref. https://github.com/jrl-umi3218/jrl-cmakemodules#supported-cmake-versions

ManifoldFR replied:
Oh yes, we've removed support for the old ones since this PR was initially created! Will fix this asap. Thanks

ManifoldFR committed:
Apply suggestions from code review by @nim65s

Co-authored-by: Guilhem Saurel <guilhem.saurel@laas.fr>
